content/wayland-rs-v0.20.markdown

```diff
@@ -0,0 +1,158 @@
+object consists in providing a type implementing the appropriate `Implementation<Meta, Msg>`
+trait.
+
+This trait has a very simple definition:
+
+```rust
+pub trait Implementation<Meta, Msg> {
+    fn receive(&mut self, msg: Msg, meta: Meta);
+}
+```
+
+Here the `Meta` type parameter represent some metadata associated with the message you
+receive (often a `Proxy`/`Resource` handle to the wayland object receiving the message),
+while `Msg` is a type representing the message itself (often an enum of the possible wayland
+messages this object can receive).
+
+The convenience is that `Implementation<Meta, Msg>` is automatically implemented for all
+type that implement `FnMut(Msg, Meta)`. As such, you can easily provide closures as
+implementations, with all the ergonomics of capturing values and code compacity that
+implies. The only catch being that the implementations must be `'static`, so you'll need
+to rely on `Arc` or `Rc` to share data. The runtime cost implied should be negligible however,
+the wayland protocol part should hardly be a bottleneck for any application.
+
+There are also two ways to implement an object, the default one requiring from the 
+implementation to be `Send`, and a secondary one that relaxes this requirement provided you
+provide a token that proves you are doing it from the same thread as the one on which the
+implementation will be invoked.
+
+## Smithay's Client Toolkit
+
+I also took the occasion of this large API rework to fuse
+[wayland-kbd](https://github.com/smithay/wayland-kbd) and
+[wayland-window](https://github.com/smithay/wayland-window) into a single crate named
+[smithay-client-toolkit](https://github.com/smithay/client-toolkit), which will be some
+kind of an equivalent of [smithay](https://github.com/smithay/smithay), but for client
+applications. Given the scope of what is possible, it'll clearly remain much simpler than
+smithay itself, as it only abstracts the wayland protocol, and nothing like all of
+smithay's backends.
+
+Currently, it only provides some barebones functionalities, but there is still a lot
+of room in this toolkit to grow if you want to get involved.
+
+As an example, a minimal client could be created like this:
+
+```rust
+extern crate smithay_client_toolkit as sctk;
+
+use sctk::Environment;
+use sctk::window::{BasicFrame, Window};
+
+// wayland-client is re-exported in the client toolkit, for convenience
+use sctk::reexports::client::{Display, Proxy};
+use sctk::reexports::client::protocol::wl_display::RequestsTrait as DisplayRequests;
+use sctk::reexports::client::protocol::wl_compositor::RequestsTrait as CompositorRequests;
+
+fn main() {
+    // Connect to a wayland server
+    let (display, mut event_queue) = Display::connect_to_env().unwrap();
+
+    // The Environment is an abstraction binding for you most of the
+    // classic globals. Otherwise this job is very boring and repetitive.
```

Comment:
[...] an abstraction that binds most of the classic globals for you. Doing this manually is very boring and repetitive.